src/Language/PureScript/CodeGen/JS.hs

```diff
@@ -397,29 +396,37 @@ moduleToJs (Module coms mn _ imps exps foreigns decls) foreign_ =
       return (AST.VariableIntroduction Nothing propVar (Just (accessorString prop (AST.Var Nothing varName))) : js)
   literalToBinderJS varName done (ArrayLiteral bs) = do
     js <- go done 0 bs
-    return [AST.IfElse Nothing (AST.Binary Nothing AST.EqualTo (accessorString "length" (AST.Var Nothing varName)) (AST.NumericLiteral Nothing (Left (fromIntegral $ length bs)))) (AST.Block Nothing js) Nothing]
+    return [AST.IfElse Nothing (AST.Binary Nothing AST.EqualTo (accessorString "length" (AST.Var Nothing varName)) (AST.NumericLiteral Nothing (LitInt (fromIntegral $ length bs)))) (AST.Block Nothing js) Nothing]
     where
     go :: [AST] -> Integer -> [Binder Ann] -> m [AST]
     go done' _ [] = return done'
     go done' index (binder:bs') = do
       elVar <- freshName
       done'' <- go done' (index + 1) bs'
       js <- binderToJs elVar done'' binder
-      return (AST.VariableIntroduction Nothing elVar (Just (AST.Indexer Nothing (AST.NumericLiteral Nothing (Left index)) (AST.Var Nothing varName))) : js)
+      return (AST.VariableIntroduction Nothing elVar (Just (AST.Indexer Nothing (AST.NumericLiteral Nothing (LitInt index)) (AST.Var Nothing varName))) : js)
 
   -- Check that all integers fall within the valid int range for JavaScript.
   checkIntegers :: AST -> m ()
   checkIntegers = void . everywhereTopDownM go
     where
     go :: AST -> m AST
-    go (AST.Unary _ AST.Negate (AST.NumericLiteral ss (Left i))) =
+    go (AST.Unary _ AST.Negate (AST.NumericLiteral ss (LitInt i))) =
       -- Move the negation inside the literal; since this is a top-down
       -- traversal doing this replacement will stop the next case from raising
       -- the error when attempting to use -2147483648, as if left unrewritten
-      -- the value is `Unary Negate (NumericLiteral (Left 2147483648))`, and
+      -- the value is `Unary Negate (NumericLiteral (LitInt 2147483648))`, and
       -- 2147483648 is larger than the maximum allowed int.
-      return $ AST.NumericLiteral ss (Left (-i))
-    go js@(AST.NumericLiteral _ (Left i)) =
+      return $ AST.NumericLiteral ss (LitInt (-i))
+    go (AST.Unary _ AST.Negate (AST.NumericLiteral _ (LitUInt i))) = do
+      let maxInt = 2 * 2147483647
+      throwError . errorMessage $ NegativeUInt (- toInteger i) "JavaScript" maxInt
```

Comment:
Do we not want there to be a `Ring UInt` instance? I'd suggest we should have it and have it wrap around - as far as I can tell, that ought to obey all the ring laws, because I think it should be isomorphic to ℤ/(2^32)ℤ as a ring (just like `Int` is). Well, provided that we gloss over the multiplication accuracy issue (https://github.com/purescript/purescript/issues/2980).

Reply:
I need to study my math more :sweat_smile: I thought `negate` was a class method on `Ring`. I'll add the relevant instances in `purescript-prelude`.

How can I setup the version of Prelude that the tests use? I want to refer to the branch I'm working on for this feature.

Reply:
I think from a mathematical perspective, it would be equally valid to put either `negate` or `sub` in the class, and then define the other one in terms of the first one. We probably decided to put `sub` in the class because it allows more sensible generated code for subtractions.

I think you can setup the version of Prelude that the tests use by modifying `tests/support/bower.json`.

Reply:
Thanks! I'll update the tests and provide the instances :smile: